File: ios/Sources/TransactPluginPlugin/TransactPluginPlugin.swift
import Foundation
import Capacitor
import AtomicTransact

@objc(TransactPluginPlugin)
public class TransactPluginPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "TransactPluginPlugin"
    public let jsName = "TransactPlugin"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "presentTransact", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "presentAction", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hideTransact", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "pauseTransact", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "resumeTransact", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "resolveDataRequest", returnType: CAPPluginReturnPromise)
    ]

    private var dataResponseHandler: (([String: Any]?) -> Void)?
    private var pausedTransactRef: Atomic.PausedTransactRef?

    // MARK: - Environment Parsing

    private func parseEnvironment(_ environmentData: [String: Any]?) -> TransactEnvironment {
        guard let env = environmentData,
              let environment = env["environment"] as? String else {
            return .production
        }

        switch environment {
        case "sandbox":
            return .sandbox
        case "custom":
            let transactPath = env["transactPath"] as? String ?? "https://transact.atomicfi.com"
            let apiPath = env["apiPath"] as? String ?? "https://api.atomicfi.com"
            return .custom(transactPath: transactPath, apiPath: apiPath)
        default:
            return .production
        }
    }

    private func parsePresentationStyle(_ style: String?) -> UIModalPresentationStyle {
        switch style {
        case "fullScreen":
            return .fullScreen
        default:
            return .formSheet
        }
    }

    // MARK: - presentTransact

    @objc func presentTransact(_ call: CAPPluginCall) {
        guard let configData = call.getObject("config") else {
            call.reject("Config is required")
            return
        }

        let environmentData = call.getObject("environment")
        let presentationStyle = call.getString("presentationStyle")
        let debug = call.getBool("debug") ?? false

        DispatchQueue.main.async { [weak self] in
            guard let self = self,
                  let source = self.bridge?.viewController else {
                call.reject("Unable to get view controller")
                return
            }

            let parsedEnvironment = self.parseEnvironment(environmentData)
            let parsedPresentationStyle = self.parsePresentationStyle(presentationStyle)

            do {
                var json = configData as [String: Any]

                // Default language to "en" if not provided
                if json["language"] == nil {
                    json["language"] = "en"
                }

                // Add platform info using SDK defaults with -capacitor suffix
                json["platform"] = AtomicConfig.Platform(suffixed: "capacitor").encode()

                guard let data = try? JSONSerialization.data(withJSONObject: json, options: []) else {
                    call.reject("Failed to serialize config")
                    return
                }

                let decoder = JSONDecoder()
                let config = try decoder.decode(AtomicConfig.self, from: data)

                Task { @MainActor in
                    await Atomic.setDebug(isEnabled: debug, forwardLogs: { [weak self] message in
                        DispatchQueue.main.async {
                            self?.notifyListeners("onDebugLog", data: ["message": message])
                        }
                    })

                    Atomic.presentTransact(
                    from: source,
                    config: config,
                    environment: parsedEnvironment,
                    presentationStyle: parsedPresentationStyle,
                    onInteraction: { [weak self] interaction in
                        self?.notifyListeners("onInteraction", data: [
                            "name": interaction.name,
                            "value": interaction.value
                        ])
                    },
                    onDataRequest: { [weak self] request async -> TransactDataResponse? in
                        return await withCheckedContinuation { continuation in
                            self?.dataResponseHandler = { responseData in
                                guard let responseDict = responseData else {
                                    continuation.resume(returning: nil)
                                    return
                                }

                                do {
                                    let jsonData = try JSONSerialization.data(withJSONObject: responseDict, options: [])
                                    let response = try JSONDecoder().decode(TransactDataResponse.self, from: jsonData)
                                    continuation.resume(returning: response)
                                } catch {
                                    continuation.resume(returning: nil)
                                }
                            }

                            // Build event data from the request
                            var eventData: [String: Any] = [:]
                            for (key, value) in request.data {
                                eventData[key] = value
                            }
                            eventData["userId"] = request.userId
                            eventData["identifier"] = request.identifier
                            eventData["fields"] = request.fields
                            if let taskId = request.taskId {
                                eventData["taskId"] = taskId
                            }

                            self?.notifyListeners("onDataRequest", data: eventData)
                        }
                    },
                    onAuthStatusUpdate: { [weak self] status in
                        self?.notifyListeners("onAuthStatusUpdate", data: status.toDictionary())
                    },
                    onTaskStatusUpdate: { [weak self] status in
                        self?.notifyListeners("onTaskStatusUpdate", data: status.toDictionary())
                    },
                    onLaunch: { [weak self] in
                        self?.notifyListeners("onLaunch", data: [:])
                    },
                    onCompletion: { [weak self] result in
                        switch result {
                        case .finished(let response):
                            let data = self?.sanitizeDictionary(response.data) ?? [:]
                            self?.notifyListeners("onFinish", data: data)
                            call.resolve(["finished": data])
                        case .closed(let response):
                            let data = self?.sanitizeDictionary(response.data) ?? [:]
                            self?.notifyListeners("onClose", data: data)
                            call.resolve(["closed": data])
                        case .error:
                            call.resolve(["error": "Unknown error"])
                        case .transactDismissed:
                            call.resolve(["closed": ["reason": "dismissed"]])
                        @unknown default:
                            call.resolve(["error": "Unknown error"])
                        }
                    }
                )
                }
            } catch let DecodingError.keyNotFound(key, context) {
                call.reject("Config error: Missing key '\(key.stringValue)' at path: \(context.codingPath.map(\.stringValue).joined(separator: ".")). Debug: \(context.debugDescription)")
            } catch let DecodingError.typeMismatch(type, context) {
                call.reject("Config error: Type mismatch for \(type) at path: \(context.codingPath.map(\.stringValue).joined(separator: ".")). Debug: \(context.debugDescription)")
            } catch let DecodingError.valueNotFound(type, context) {
                call.reject("Config error: Value not found for \(type) at path: \(context.codingPath.map(\.stringValue).joined(separator: ".")). Debug: \(context.debugDescription)")
            } catch {
                call.reject("Config error: \(String(describing: error))")
            }
        }
    }

    // MARK: - presentAction

    @objc func presentAction(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else {
            call.reject("id is required")
            return
        }

        let environmentData = call.getObject("environment")
        let presentationStyle = call.getString("presentationStyle")
        let debug = call.getBool("debug") ?? false

        DispatchQueue.main.async { [weak self] in
            guard let self = self,
                  let source = self.bridge?.viewController else {
                call.reject("Unable to get view controller")
                return
            }

            let parsedEnvironment = self.parseEnvironment(environmentData)
            let parsedPresentationStyle = self.parsePresentationStyle(presentationStyle)

            // Parse optional theme
            var theme = AtomicConfig.Theme()
            if let themeData = call.getObject("theme") {
                if let jsonData = try? JSONSerialization.data(withJSONObject: themeData, options: []),
                   let parsedTheme = try? JSONDecoder().decode(AtomicConfig.Theme.self, from: jsonData) {
                    theme = parsedTheme
                }
            }

            // Parse optional metadata
            var metadata: [String: String]? = nil
            if let metadataData = call.getObject("metadata") {
                metadata = metadataData.compactMapValues { $0 as? String }
            }

            Task { @MainActor in
                await Atomic.setDebug(isEnabled: debug, forwardLogs: { [weak self] message in
                    DispatchQueue.main.async {
                        self?.notifyListeners("onDebugLog", data: ["message": message])
                    }
                })

                Atomic.presentAction(
                from: source,
                id: id,
                environment: parsedEnvironment,
                presentationStyle: parsedPresentationStyle,
                theme: theme,
                metadata: metadata,
                onLaunch: { [weak self] in
                    self?.notifyListeners("onLaunch", data: [:])
                },
                onAuthStatusUpdate: { [weak self] status in
                    self?.notifyListeners("onAuthStatusUpdate", data: status.toDictionary())
                },
                onTaskStatusUpdate: { [weak self] status in
                    self?.notifyListeners("onTaskStatusUpdate", data: status.toDictionary())
                },
                onCompletion: { [weak self] result in
                    switch result {
                    case .finished(let response):
                        let data = self?.sanitizeDictionary(response.data) ?? [:]
                        self?.notifyListeners("onFinish", data: data)
                        call.resolve(["finished": data])
                    case .closed(let response):
                        let data = self?.sanitizeDictionary(response.data) ?? [:]
                        self?.notifyListeners("onClose", data: data)
                        call.resolve(["closed": data])
                    case .error:
                        call.resolve(["error": "Unknown error"])
                    case .transactDismissed:
                        call.resolve(["closed": ["reason": "dismissed"]])
                    @unknown default:
                        call.resolve(["error": "Unknown error"])
                    }
                }
            )
            }
        }
    }

    // MARK: - hideTransact

    @objc func hideTransact(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Atomic.hideTransact()
            call.resolve()
        }
    }

    // MARK: - pauseTransact

    @objc func pauseTransact(_ call: CAPPluginCall) {
        let animated = call.getBool("animated") ?? true

        Task { @MainActor [weak self] in
            guard let self = self else {
                call.reject("Plugin deallocated")
                return
            }

            do {
                let ref = try await Atomic.pauseTransact(animated: animated)
                self.pausedTransactRef = ref
                call.resolve()
            } catch Atomic.PauseTransactError.transactNotPresented {
                call.reject("Transact is not currently presented")
            } catch {
                call.reject("Failed to pause Transact: \(String(describing: error))")
            }
        }
    }

    // MARK: - resumeTransact

    @objc func resumeTransact(_ call: CAPPluginCall) {
        let animated = call.getBool("animated") ?? true

        DispatchQueue.main.async { [weak self] in
            guard let self = self else {
                call.reject("Plugin deallocated")
                return
            }
            guard let ref = self.pausedTransactRef else {
                call.reject("No paused Transact session to resume")
                return
            }
            guard let source = self.bridge?.viewController else {
                call.reject("Unable to get view controller")
                return
            }

            ref.resume(source: source, animated: animated)
            self.pausedTransactRef = nil
            call.resolve()
        }
    }

    // MARK: - resolveDataRequest

    @objc func resolveDataRequest(_ call: CAPPluginCall) {
        guard let handler = dataResponseHandler else {
            call.reject("No active data request")
            return
        }

        var responseData: [String: Any] = [:]
        if let card = call.getObject("card") {
            responseData["card"] = card
        }
        if let identity = call.getObject("identity") {
            responseData["identity"] = identity
        }

        handler(responseData.isEmpty ? nil : responseData)
        dataResponseHandler = nil
        call.resolve()
    }

    // MARK: - Helpers

    /// Ensures all values in the dictionary are JSON-serializable for Capacitor
    private func sanitizeDictionary(_ dict: [String: Any]) -> [String: Any] {
        var result: [String: Any] = [:]
        for (key, value) in dict {
            if let nested = value as? [String: Any] {
                result[key] = sanitizeDictionary(nested)
            } else if let array = value as? [Any] {
                result[key] = array
            } else if value is String || value is Int || value is Double || value is Bool {
                result[key] = value
            } else if let number = value as? NSNumber {
                result[key] = number
            } else {
                result[key] = "\(value)"
            }
        }
        return result
    }
}

// MARK: - Serialization Extensions

extension TransactCompany {
    func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [
            "_id": id,
            "name": name
        ]
        if let branding = branding {
            var brandingDict: [String: Any] = [
                "color": branding.color
            ]
            var logoDict: [String: Any] = [
                "url": branding.logo.url
            ]
            if let bgColor = branding.logo.backgroundColor {
                logoDict["backgroundColor"] = bgColor
            }
            brandingDict["logo"] = logoDict
            dict["branding"] = brandingDict
        }
        return dict
    }
}

extension TransactAuthStatusUpdate {
    func toDictionary() -> [String: Any] {
        return [
            "status": status.rawValue,
            "company": company.toDictionary()
        ]
    }
}

extension TransactTaskStatusUpdate {
    func toDictionary() -> [String: Any] {
        var result: [String: Any] = [
            "taskId": taskId,
            "product": product.rawValue,
            "status": status.rawValue,
            "company": company.toDictionary()
        ]

        if let failReason = failReason {
            result["failReason"] = failReason
        }

        if let switchData = switchData {
            var switchMap: [String: Any] = [:]
            let payment = switchData.paymentMethod
            var paymentMap: [String: Any] = [
                "id": payment.id,
                "title": payment.title,
                "type": payment.type.rawValue
            ]

            if let expiry = payment.expiry { paymentMap["expiry"] = expiry }
            if let brand = payment.brand { paymentMap["brand"] = brand }
            if let lastFour = payment.lastFour { paymentMap["lastFour"] = lastFour }
            if let routingNumber = payment.routingNumber { paymentMap["routingNumber"] = routingNumber }
            if let accountType = payment.accountType { paymentMap["accountType"] = accountType }
            if let lastFourAccount = payment.lastFourAccountNumber { paymentMap["lastFourAccountNumber"] = lastFourAccount }

            switchMap["paymentMethod"] = paymentMap
            result["switchData"] = switchMap
        }

        if let depositData = depositData {
            var depositMap: [String: Any] = [:]
            if let accountType = depositData.accountType { depositMap["accountType"] = accountType }
            if let distributionAmount = depositData.distributionAmount { depositMap["distributionAmount"] = distributionAmount }
            if let distributionType = depositData.distributionType { depositMap["distributionType"] = "\(distributionType)" }
            if let lastFour = depositData.lastFour { depositMap["lastFour"] = lastFour }
            if let routingNumber = depositData.routingNumber { depositMap["routingNumber"] = routingNumber }
            if let title = depositData.title { depositMap["title"] = title }
            result["depositData"] = depositMap
        }

        if let managedBy = managedBy {
            result["managedBy"] = ["company": managedBy.company.toDictionary()]
        }

        return result
    }
}

File: android/src/main/java/com/atomicfi/transact/TransactPluginPlugin.kt
package com.atomicfi.transact

import android.util.Base64
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import financial.atomic.transact.ActionConfig
import financial.atomic.transact.Config
import financial.atomic.transact.PausedTransactRef
import financial.atomic.transact.Transact
import financial.atomic.transact.receiver.TransactBroadcastReceiver
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import org.json.JSONObject

@CapacitorPlugin(name = "TransactPlugin")
class TransactPluginPlugin : Plugin() {

    private var savedCall: PluginCall? = null
    private var pausedRef: PausedTransactRef? = null
    private val pluginScope = CoroutineScope(Dispatchers.Main + Job())

    // The `environment` option accepts a shorthand string ("production" | "sandbox") or a
    // full object ({ environment, transactPath?, apiPath? }). Normalize both to a name and
    // optional transactPath.
    private fun readEnvironment(call: PluginCall): Pair<String, String?> {
        // Shorthand string form.
        call.getString("environment")?.let { return Pair(it, null) }

        // Object form.
        val environmentObj = call.getObject("environment") ?: return Pair("production", null)
        val name = environmentObj.optString("environment", "production")
        val transactPath = if (environmentObj.has("transactPath")) {
            environmentObj.optString("transactPath")
        } else {
            null
        }
        return Pair(name, transactPath)
    }

    private fun parseEnvironmentURL(name: String, transactPath: String?): String {
        return when (name) {
            "sandbox" -> "https://transact-sandbox.atomicfi.com"
            "custom" -> transactPath ?: "https://transact.atomicfi.com"
            else -> "https://transact.atomicfi.com"
        }
    }

    private fun parseActionEnvironment(name: String, transactPath: String?): Pair<Config.Environment, String?> {
        return when (name) {
            "sandbox" -> Pair(Config.Environment.SANDBOX, null)
            "custom" -> Pair(Config.Environment.CUSTOM, transactPath ?: "https://transact.atomicfi.com")
            else -> Pair(Config.Environment.PRODUCTION, null)
        }
    }

    @PluginMethod
    fun presentTransact(call: PluginCall) {
        val configObj = call.getObject("config")
        if (configObj == null) {
            call.reject("Config is required")
            return
        }

        val (environmentName, transactPath) = readEnvironment(call)
        val debug = call.getBoolean("debug") ?: false
        val wrapperVersion = call.getString("wrapperVersion") ?: ""
        val activity = bridge.activity
        if (activity == null) {
            call.reject("Activity not available")
            return
        }

        savedCall = call

        // Add platform info using SDK helper with `capacitor-<wrapper version>` suffix.
        val platformMap = Config.Platform.suffixed("capacitor-$wrapperVersion").encode()
        configObj.put("platform", JSONObject(platformMap as Map<String, Any?>))

        // Base64 encode the config for the token-based constructor
        val token = Base64.encodeToString(
            configObj.toString().toByteArray(Charsets.UTF_8),
            Base64.NO_WRAP
        )

        val environmentURL = parseEnvironmentURL(environmentName, transactPath)

        val config = Config(
            token = token,
            environment = "CUSTOM",
            environmentURL = environmentURL,
            webContentsDebuggingEnabled = debug
        )

        activity.runOnUiThread {
            try {
                Transact.present(activity, config, object : TransactBroadcastReceiver() {
                    override fun onClose(data: JSONObject) {
                        notifyListeners("onClose", jsonToJSObject(data))

                        val result = JSObject()
                        result.put("closed", jsonToJSObject(data))
                        savedCall?.resolve(result)
                        savedCall = null
                    }

                    override fun onFinish(data: JSONObject) {
                        notifyListeners("onFinish", jsonToJSObject(data))

                        val result = JSObject()
                        result.put("finished", jsonToJSObject(data))
                        savedCall?.resolve(result)
                        savedCall = null
                    }

                    override fun onInteraction(data: JSONObject) {
                        notifyListeners("onInteraction", jsonToJSObject(data))
                    }

                    override fun onDataRequest(data: JSONObject) {
                        notifyListeners("onDataRequest", jsonToJSObject(data))
                    }

                    override fun onAuthStatusUpdate(data: JSONObject) {
                        notifyListeners("onAuthStatusUpdate", jsonToJSObject(data))
                    }

                    override fun onTaskStatusUpdate(data: JSONObject) {
                        notifyListeners("onTaskStatusUpdate", jsonToJSObject(data))
                    }
                })
            } catch (e: Exception) {
                call.reject("Failed to present Transact", e)
                savedCall = null
            }
        }
    }

    @PluginMethod
    fun presentAction(call: PluginCall) {
        val id = call.getString("id")
        if (id == null) {
            call.reject("id is required")
            return
        }

        val (environmentName, transactPath) = readEnvironment(call)
        val debug = call.getBoolean("debug") ?: false
        val wrapperVersion = call.getString("wrapperVersion") ?: ""
        val activity = bridge.activity
        if (activity == null) {
            call.reject("Activity not available")
            return
        }

        savedCall = call

        val (env, envURL) = parseActionEnvironment(environmentName, transactPath)

        val config = ActionConfig(
            id = id,
            environment = env,
            environmentURL = envURL,
            webContentsDebuggingEnabled = debug
        )
        config.platform = Config.Platform.suffixed("capacitor-$wrapperVersion")

        activity.runOnUiThread {
            try {
                Transact.presentAction(activity, config)

                Transact.registerReceiver(activity, object : TransactBroadcastReceiver() {
                    override fun onLaunch() {
                        notifyListeners("onLaunch", JSObject())
                    }

                    override fun onClose(data: JSONObject) {
                        notifyListeners("onClose", jsonToJSObject(data))

                        val result = JSObject()
                        result.put("closed", jsonToJSObject(data))
                        savedCall?.resolve(result)
                        savedCall = null
                    }

                    override fun onFinish(data: JSONObject) {
                        notifyListeners("onFinish", jsonToJSObject(data))

                        val result = JSObject()
                        result.put("finished", jsonToJSObject(data))
                        savedCall?.resolve(result)
                        savedCall = null
                    }

                    override fun onAuthStatusUpdate(data: JSONObject) {
                        notifyListeners("onAuthStatusUpdate", jsonToJSObject(data))
                    }

                    override fun onTaskStatusUpdate(data: JSONObject) {
                        notifyListeners("onTaskStatusUpdate", jsonToJSObject(data))
                    }
                })
            } catch (e: Exception) {
                call.reject("Failed to present action", e)
                savedCall = null
            }
        }
    }

    @PluginMethod
    fun hideTransact(call: PluginCall) {
        val activity = bridge.activity
        if (activity == null) {
            call.reject("Activity not available")
            return
        }

        activity.runOnUiThread {
            try {
                Transact.hideTransact(activity)
                call.resolve()
            } catch (e: Exception) {
                call.reject("Failed to hide Transact", e)
            }
        }
    }

    @PluginMethod
    fun pauseTransact(call: PluginCall) {
        val animated = call.getBoolean("animated") ?: true

        pluginScope.launch {
            try {
                val ref = Transact.pauseTransact(animated)
                pausedRef = ref
                call.resolve()
            } catch (e: Exception) {
                call.reject("Failed to pause Transact", e)
            }
        }
    }

    @PluginMethod
    fun resumeTransact(call: PluginCall) {
        val animated = call.getBoolean("animated") ?: true
        val activity = bridge.activity
        if (activity == null) {
            call.reject("Activity not available")
            return
        }
        val ref = pausedRef
        if (ref == null) {
            call.reject("No paused Transact session to resume")
            return
        }

        activity.runOnUiThread {
            try {
                ref.resume(activity, animated)
                pausedRef = null
                call.resolve()
            } catch (e: Exception) {
                call.reject("Failed to resume Transact", e)
            }
        }
    }

    @PluginMethod
    fun resolveDataRequest(call: PluginCall) {
        val activity = bridge.activity
        if (activity == null) {
            call.reject("Activity not available")
            return
        }

        val cardObj = call.getObject("card")
        val identityObj = call.getObject("identity")

        var card: Config.TransactDataResponse.CardData? = null
        if (cardObj != null) {
            val number = cardObj.getString("number")
            if (number == null) {
                call.reject("Card number is required")
                return
            }
            card = Config.TransactDataResponse.CardData(
                number = number,
                expiry = cardObj.optStringOrNull("expiry"),
                cvv = cardObj.optStringOrNull("cvv")
            )
        }

        var identity: Config.TransactDataResponse.Identity? = null
        if (identityObj != null) {
            identity = Config.TransactDataResponse.Identity(
                firstName = identityObj.optStringOrNull("firstName"),
                lastName = identityObj.optStringOrNull("lastName"),
                postalCode = identityObj.optStringOrNull("postalCode"),
                address = identityObj.optStringOrNull("address"),
                address2 = identityObj.optStringOrNull("address2"),
                city = identityObj.optStringOrNull("city"),
                state = identityObj.optStringOrNull("state"),
                phone = identityObj.optStringOrNull("phone"),
                email = identityObj.optStringOrNull("email")
            )
        }

        val response = Config.TransactDataResponse(card = card, identity = identity)

        activity.runOnUiThread {
            try {
                Transact.sendData(activity, response)
                call.resolve()
            } catch (e: Exception) {
                call.reject("Failed to send data response", e)
            }
        }
    }

    private fun jsonToJSObject(json: JSONObject): JSObject {
        return try {
            JSObject(json.toString())
        } catch (e: Exception) {
            JSObject()
        }
    }

    private fun JSONObject.optStringOrNull(key: String): String? {
        return if (has(key) && !isNull(key)) getString(key) else null
    }
}
